File: TSCompiler/JavaRTL/String.java
package JavaRTL;

public class String extends Any {

    public java.lang.String _value = "";

    public String() {}

    public String(int value){
        this._value = java.lang.String.valueOf(value);
    }

    public String(double value){
        this._value = java.lang.String.valueOf(value);
    }

    public String(java.lang.String value){
        this._value = value;
    }

    public String(JavaRTL.Any value){
        if (value == null) {
            value = new JavaRTL.Null();
        }
        this._value = value.toString();
    }

    public String plus(JavaRTL.Any other) {
        return new String(this._value + (new String(other))._value);
    }

    public JavaRTL.Boolean less(String other) {
        boolean result = this._value.compareTo(other._value) < 0;
        return new JavaRTL.Boolean(result);
    }
    public JavaRTL.Boolean less(JavaRTL.Null other) {
        return (new JavaRTL.Number(this)).less(new Number(other));
    }
    public JavaRTL.Boolean less(JavaRTL.Undefined other) {
        return new JavaRTL.Boolean(false);
    }

    // that method not adding into ASI
    public JavaRTL.Boolean less(Any other) {
        if (other instanceof JavaRTL.Array) {
            other = new String(other);
        }
        if (other instanceof String) {
            boolean result = this._value.compareTo(((String)other)._value) < 0;
            return new JavaRTL.Boolean(result);
        }
        return super.less(other);       
    }

    public JavaRTL.Boolean great(String other) {
        boolean result = this._value.compareTo(other._value) > 0;
        return new JavaRTL.Boolean(result);
    }
    public JavaRTL.Boolean great(JavaRTL.Null other) {
        return (new JavaRTL.Number(this)).great(new Number(other));
    }
    public JavaRTL.Boolean great(JavaRTL.Undefined other) {
        return new JavaRTL.Boolean(false);
    }

    // that method not adding into ASI
    public JavaRTL.Boolean great(Any other) {
        if (other instanceof JavaRTL.Array) {
            other = new String(other);
        }
        if (other instanceof String) {
            boolean result = this._value.compareTo(((String)other)._value) > 0;
            return new JavaRTL.Boolean(result);
        }
        return super.great(other);       
    }

    public JavaRTL.Boolean equal(String other) {
        boolean result = this._value.compareTo(other._value) == 0;
        return new JavaRTL.Boolean(result);
    }
    public JavaRTL.Boolean equal(JavaRTL.Null other) {
        return new JavaRTL.Boolean(false);
    }
    public JavaRTL.Boolean equal(JavaRTL.Undefined other) {
        return new JavaRTL.Boolean(false);
    }

    // that method not adding into ASI
    public JavaRTL.Boolean equal(Any other) {
        if (other instanceof JavaRTL.Array) {
            other = new String(other);
        }
        if (other instanceof String) {
            boolean result = this._value.compareTo(((String)other)._value) == 0;
            return new JavaRTL.Boolean(result);
        }
        if (other instanceof JavaRTL.Void || other instanceof JavaRTL.Null) {
            return new JavaRTL.Boolean(false);
        }
        return super.equal(other);       
    }

    public JavaRTL.Boolean notEqual(String other) {
        boolean result = this._value.compareTo(other._value) != 0;
        return new JavaRTL.Boolean(result);
    }
    public JavaRTL.Boolean notEqual(JavaRTL.Null other) {
        return new JavaRTL.Boolean(true);
    }
    public JavaRTL.Boolean notEqual(JavaRTL.Undefined other) {
        return new JavaRTL.Boolean(true);
    }

    // that method not adding into ASI
    public JavaRTL.Boolean notEqual(Any other) {
        if (other instanceof JavaRTL.Array) {
            other = new String(other);
        }
        if (other instanceof String) {
            boolean result = this._value.compareTo(((String)other)._value) != 0;
            return new JavaRTL.Boolean(result);
        }
        if (other instanceof JavaRTL.Void || other instanceof JavaRTL.Null) {
            return new JavaRTL.Boolean(true);
        }
        return super.notEqual(other);       
    }

    public JavaRTL.Boolean lessEqual(String other) {
        boolean result = this._value.compareTo(other._value) <= 0;
        return new JavaRTL.Boolean(result);
    }
    public JavaRTL.Boolean lessEqual(JavaRTL.Null other) {
        return (new JavaRTL.Number(this)).lessEqual(new Number(other));
    }
    public JavaRTL.Boolean lessEqual(JavaRTL.Undefined other) {
        return new JavaRTL.Boolean(false);
    }

    // that method not adding into ASI
    public JavaRTL.Boolean lessEqual(Any other) {
        if (other instanceof JavaRTL.Array) {
            other = new String(other);
        }
        if (other instanceof String) {
            boolean result = this._value.compareTo(((String)other)._value) <= 0;
            return new JavaRTL.Boolean(result);
        }
        return super.lessEqual(other);       
    }

    public JavaRTL.Boolean greatEqual(String other) {
        boolean result = this._value.compareTo(other._value) >= 0;
        return new JavaRTL.Boolean(result);
    }
    public JavaRTL.Boolean greatEqual(JavaRTL.Null other) {
        return (new JavaRTL.Number(this)).greatEqual(new Number(other));
    }
    public JavaRTL.Boolean greatEqual(JavaRTL.Undefined other) {
        return new JavaRTL.Boolean(false);
    }

    // that method not adding into ASI
    public JavaRTL.Boolean greatEqual(Any other) {
        if (other instanceof JavaRTL.Array) {
            other = new String(other);
        }
        if (other instanceof String) {
            boolean result = this._value.compareTo(((String)other)._value) >= 0;
            return new JavaRTL.Boolean(result);
        }
        return super.greatEqual(other);       
    }

    @Override
    public java.lang.String toString() {
        return this._value;
    }
}

File: TSCompiler/Semantic/RTL/string.cpp
#include "../semantic.h"

ClassDeclarationNode *Semantic::createStringClass()
{
    auto *body = ClassElementListNode::makeEmpty();

    // Class constructors
    {
        auto *clsConstructor = new ClassElementNode(
            RequiredParameterListNode::makeEmpty(), nullptr);
        clsConstructor->analyzeArguments();
        body->add(clsConstructor);
        clsConstructor->propertyAndReturnType =
            new TypeNode(new JvmDataType(JvmDataType::Type::Void));
    }

    std::vector<JvmDataType> types = {JvmDataType(JvmDataType::Type::Double),
                                      JvmDataType(JvmDataType::Type::Int),
                                      JvmDataType(JvmDataType::Type::String),
                                      RTL_ANY_TYPE};

    for (const auto &jvmType : types)
    {
        auto *type = new TypeNode(new JvmDataType(jvmType));
        auto *param = new RequiredParameterNode("value", type);
        auto *clsConstructor =
            new ClassElementNode(new RequiredParameterListNode(param), nullptr);
        clsConstructor->analyzeArguments();
        body->add(clsConstructor);
        clsConstructor->propertyAndReturnType =
            new TypeNode(new JvmDataType(JvmDataType::Type::Void));
    }

    // Methods

    // Plus

    {
        auto *type = new TypeNode(new JvmDataType(RTL_ANY_TYPE));
        auto *param = new RequiredParameterNode("other", type);
        auto *method = new ClassElementNode(
            "plus", new RequiredParameterListNode(param),
            new TypeNode(new JvmDataType(RTL_STRING_TYPE)), nullptr);
        method->analyzeArguments();
        body->add(method);
    }

    {
        auto *method = new ClassElementNode(
            "uPlus", RequiredParameterListNode::makeEmpty(),
            new TypeNode(new JvmDataType(RTL_NUMBER_TYPE)), nullptr);
        body->add(method);
    }
    {
        auto *method = new ClassElementNode(
            "uMinus", RequiredParameterListNode::makeEmpty(),
            new TypeNode(new JvmDataType(RTL_NUMBER_TYPE)), nullptr);
        body->add(method);
    }

    // Less

    std::vector<JvmDataType> operationTypes = {RTL_STRING_TYPE, RTL_NULL_TYPE,
                                               RTL_UNDEFINED_TYPE};

    for (const auto &jvmType : operationTypes)
    {
        auto *type = new TypeNode(new JvmDataType(jvmType));
        auto *param = new RequiredParameterNode("other", type);
        auto *method = new ClassElementNode(
            "less", new RequiredParameterListNode(param),
            new TypeNode(new JvmDataType(RTL_BOOLEAN_TYPE)), nullptr);
        method->analyzeArguments();
        body->add(method);
    }

    // Great

    for (const auto &jvmType : operationTypes)
    {
        auto *type = new TypeNode(new JvmDataType(jvmType));
        auto *param = new RequiredParameterNode("other", type);
        auto *method = new ClassElementNode(
            "great", new RequiredParameterListNode(param),
            new TypeNode(new JvmDataType(RTL_BOOLEAN_TYPE)), nullptr);
        method->analyzeArguments();
        body->add(method);
    }

    // Equal

    for (const auto &jvmType : operationTypes)
    {
        auto *type = new TypeNode(new JvmDataType(jvmType));
        auto *param = new RequiredParameterNode("other", type);
        auto *method = new ClassElementNode(
            "equal", new RequiredParameterListNode(param),
            new TypeNode(new JvmDataType(RTL_BOOLEAN_TYPE)), nullptr);
        method->analyzeArguments();
        body->add(method);
    }

    // NotEqual

    for (const auto &jvmType : operationTypes)
    {
        auto *type = new TypeNode(new JvmDataType(jvmType));
        auto *param = new RequiredParameterNode("other", type);
        auto *method = new ClassElementNode(
            "notEqual", new RequiredParameterListNode(param),
            new TypeNode(new JvmDataType(RTL_BOOLEAN_TYPE)), nullptr);
        method->analyzeArguments();
        body->add(method);
    }

    // LessEqual

    for (const auto &jvmType : operationTypes)
    {
        auto *type = new TypeNode(new JvmDataType(jvmType));
        auto *param = new RequiredParameterNode("other", type);
        auto *method = new ClassElementNode(
            "lessEqual", new RequiredParameterListNode(param),
            new TypeNode(new JvmDataType(RTL_BOOLEAN_TYPE)), nullptr);
        method->analyzeArguments();
        body->add(method);
    }

    // GreatEqual

    for (const auto &jvmType : operationTypes)
    {
        auto *type = new TypeNode(new JvmDataType(jvmType));
        auto *param = new RequiredParameterNode("other", type);
        auto *method = new ClassElementNode(
            "greatEqual", new RequiredParameterListNode(param),
            new TypeNode(new JvmDataType(RTL_BOOLEAN_TYPE)), nullptr);
        method->analyzeArguments();
        body->add(method);
    }

    auto *rtlClass = new ClassDeclarationNode("JavaRTL/String", body);
    for (auto *elem : body->GetSeq()) elem->elemClass = rtlClass;
    return rtlClass;
}
